src/librustc_mir/interpret/snapshot.rs

```diff
@@ -46,361 +38,338 @@ pub(crate) struct InfiniteLoopDetector<'mir, 'tcx> {
+            .filter_map(|local| local.operand())
+            .flat_map(|op| op.alloc_ids().into_iter());
+
+        memory::DepthFirstSearch::with_roots(self.memory, stack_roots)
     }
 }
 
-#[derive(Eq, PartialEq)]
-struct AllocationSnapshot<'a> {
-    bytes: &'a [u8],
-    relocations: Relocations<(), AllocIdSnapshot<'a>>,
-    undef_mask: &'a UndefMask,
-    align: &'a Align,
-    size: &'a Size,
-    mutability: &'a Mutability,
-}
+impl PartialEq for InterpRef<'_, 'mir, 'tcx> {
+    fn eq(&self, other: &Self) -> bool {
+        let is_stack_eq = self.stack.iter().map(AllocIdInvariant)
+            .eq(other.stack.iter().map(AllocIdInvariant));
 
-impl<'a, Ctx> Snapshot<'a, Ctx> for &'a Allocation
-    where Ctx: SnapshotContext<'a>,
-{
-    type Item = AllocationSnapshot<'a>;
+        if !is_stack_eq {
+            return false;
+        }
 
-    fn snapshot(&self, ctx: &'a Ctx) -> Self::Item {
-        let Allocation {
-            size,
-            align,
-            mutability,
-            extra: (),
-            ..
-        } = self;
+        let live_allocs = self.live_allocs();
+        let other_live_allocs = other.live_allocs();
 
-        let all_bytes = 0..self.len();
-        // This 'inspect' is okay since following access respects undef and relocations. This does
-        // influence interpreter exeuction, but only to detect the error of cycles in evalution
-        // dependencies.
-        let bytes = self.inspect_with_undef_and_ptr_outside_interpreter(all_bytes);
+        live_allocs.map(|(_, alloc)| alloc.ok().map(AllocIdInvariant))
+            .eq(other_live_allocs.map(|(_, alloc)| alloc.ok().map(AllocIdInvariant)))
+    }
+}
 
-        let undef_mask = self.undef_mask();
-        let relocations = self.relocations();
+impl Hash for InterpRef<'_, 'mir, 'tcx> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        for frame in self.stack {
+            AllocIdInvariant(frame).hash(state);
+        }
 
-        AllocationSnapshot {
-            bytes,
-            undef_mask,
-            align,
-            size,
-            mutability,
-            relocations: relocations.snapshot(ctx),
+        for (_, alloc) in self.live_allocs() {
+            alloc.ok().map(AllocIdInvariant).hash(state);
         }
     }
 }
 
-#[derive(Eq, PartialEq)]
-struct FrameSnapshot<'a, 'tcx> {
-    instance: ty::Instance<'tcx>,
-    span: Span,
-    return_to_block: &'a StackPopCleanup,
-    return_place: Option<Place<(), AllocIdSnapshot<'a>>>,
-    locals: IndexVec<mir::Local, LocalValue<(), AllocIdSnapshot<'a>>>,
-    block: Option<mir::BasicBlock>,
-    stmt: usize,
-}
-
-impl<'a, 'mir, 'tcx, Ctx> Snapshot<'a, Ctx> for &'a Frame<'mir, 'tcx>
-    where Ctx: SnapshotContext<'a>,
-{
-    type Item = FrameSnapshot<'a, 'tcx>;
+/// A wrapper type for the various interpreter data structures that compares them without
+/// considering the numeric value of `AllocId`s.
+struct AllocIdInvariant<'a, T>(&'a T);
```

Comment:
"invariant" is an odd term here, that indicates that something doesn't change.

`IgnoreAllocId` sounds more accurate? But I have no idea why it even makes any sense to do this.^^

Reply:
Yeah, I would not be opposed to reverting to the `AllocId`-sensitive method, or removing this entirely. It is pretty cool when you accidentally omit a `break` and you're told there's a provably infinite loop in your program, and I have some sentimental connection to this code since it was my first real rust PR. However, I think a configurable instruction limit (with a low default) coupled with a real lint would have basically the same effect.

Once loops are implemented, people will actually start hitting bugs/performance issues in this code. I wanted to make sure I understood what's going on now. I'm worried people will see long-running constant evaluations grind to a halt because of this machinery and have no way to turn it off.

cc @oli-obk